presto-druid/src/main/java/io/prestosql/plugin/druid/aggregate/SingleInputAggregateFunction.java

```diff
@@ -0,0 +1,143 @@
+import io.prestosql.matching.Pattern;
+import io.prestosql.plugin.jdbc.JdbcColumnHandle;
+import io.prestosql.plugin.jdbc.JdbcExpression;
+import io.prestosql.plugin.jdbc.JdbcTypeHandle;
+import io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns;
+import io.prestosql.plugin.jdbc.expression.AggregateFunctionRule;
+import io.prestosql.spi.connector.AggregateFunction;
+import io.prestosql.spi.expression.Variable;
+import io.prestosql.spi.type.Type;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Verify.verifyNotNull;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.basicAggregation;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.expressionType;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.functionName;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.singleInput;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.variable;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SingleInputAggregateFunction
+        implements AggregateFunctionRule
+{
+    private static final Capture<Variable> INPUT = newCapture();
+    private final String prestoName; // presto aggregate function name to match
+    private final Optional<String> druidName; // mapping to druid name if its not same
+    private final Optional<JdbcTypeHandle> jdbcTypeHandle; // type handle if its different from input column
+    private Set<Type> inputTypes; // empty if all input types should match, or the set to match
+    private Optional<Type> outputType; // provide if the pattern should only match specific output type
+
+    public SingleInputAggregateFunction(
+            String prestoName,
+            Optional<String> druidName,
+            Optional<JdbcTypeHandle> jdbcTypeHandle,
+            Set<Type> inputTypes,
+            Optional<Type> outputType)
+    {
+        this.prestoName = requireNonNull(prestoName, "prestoName is null");
+        this.druidName = requireNonNull(druidName, "druidName is null");
+        this.jdbcTypeHandle = requireNonNull(jdbcTypeHandle, "jdbcTypeHandle is null");
+        this.inputTypes = requireNonNull(inputTypes, "inputTypes is null");
+        this.outputType = requireNonNull(outputType, "outputType is null");
+    }
+
+    @Override
+    public Pattern<AggregateFunction> getPattern()
+    {
+        Pattern<AggregateFunction> basePattern = basicAggregation()
+                .with(functionName().equalTo(prestoName));
+        Pattern<AggregateFunction> pattern = outputType
+                .map(type -> basePattern.with(AggregateFunctionPatterns.outputType().equalTo(type)))
+                .orElse(basePattern);
+
+        if (inputTypes.isEmpty()) {
+            return pattern.with(singleInput().matching(variable().capturedAs(INPUT)));
```

Comment:
I do not like the idea of a pushdown based on name only, without checking types.
Currently you use it only for `approx_distinct` pushdown, so let me use this as an example.

- you need to cover all types supported by the Druid connector in your tests to ensure
  Druid actually accepts `approx_distinct` on all these types
- if a new type is added to supported types, the code may break for users, without us knowing,
  unless we make some provisions to _ensure_ all supported types are test-covered

This is an example only. I am not so much concerned about `approx_distinct` (it seems easy to
support this in Druid in a generic manner, just like in Presto). However, once we have such an API
it is very tempting to use this for other functions as well, like `min`, `max`, `count` or `count(DISTINCT x)`.

Reply:
Done. Both input and output types are now required.